### PR TITLE
Type compatibility error message, when reading from .cfg

### LINKED ISSUE #2750: Unexpected exception
Reported by Stephan Merz.

## Input specification

```
--------------------------- MODULE BlockingQueue ---------------------------
(***************************************************************************)
(* Original problem and spec by Michel Charpentier                         *)
(* http://www.cs.unh.edu/~charpov/programming-tlabuffer.html               *)
(***************************************************************************)
EXTENDS Naturals, Sequences

CONSTANTS 
    \* @type: Set(PROC);
    Producers,   (* the (nonempty) set of producers                       *)
    \* @type: Set(PROC);
    Consumers,   (* the (nonempty) set of consumers                       *)
    \* @type: Int;
    BufCapacity  (* the maximum number of messages in the bounded buffer  *)

ASSUME Assumption ==
       /\ Producers # {}                      (* at least one producer *)
       /\ Consumers # {}                      (* at least one consumer *)
       /\ Producers \intersect Consumers = {} (* no thread is both consumer and producer *)
       /\ BufCapacity \in (Nat \ {0})         (* buffer capacity is at least 1 *)

-----------------------------------------------------------------------------
(***************************************************************************)
(* A PlusCal algorithm appears as a comment inside a TLA+ module. In order *)
(* to run the model checker, you first need to generate the corresponding  *)
(* TLA+ specification. If you use the TLA+ Toolbox, this is done via       *)
(* File -> Translate PlusCal Algorithm. If you use the VSCode Extension,   *)
(* run the command "TLA+: Parse module".                                   *)
(***************************************************************************)
(*
--algorithm BlockingQueue {
    variables buffer = << >>, waitP = {}, waitC = {};

    define {
        isfull(b) == Len(b) = BufCapacity
        isempty(b) == Len(b) = 0
    }

    macro wait(id) {
        if (id \in Producers) {
            waitP := waitP \union {id}
        }
        else {
            waitC := waitC \union {id};
        }
    }

    macro notify() {
        if (waitSet # {}) {
            with (t \in waitSet) {
                waitSet := waitSet \ {t};
            }
        }
    }

    macro notifyC() {
        if (waitC # {}) {
            with (t \in waitC) {
                waitC := waitC \ {t};
            }
        }
    }

    macro notifyP() {
        if (waitP # {}) {
            with (t \in waitP) {
                waitP := waitP \ {t};
            }
        }
    }

    macro notifyAll() {
        waitSet := {};
    }

    macro put(id) {
        if (isfull(buffer)) {
            wait(id);
        }
        else {
            buffer := Append(buffer, id);
            notifyC();
        }
    }

    macro get(id) {
        if (isempty(buffer)) {
            wait(id);
        }
        else {
            buffer := Tail(buffer);
            notifyP();
        }
    }

    process (p \in Producers) {
p0:     while (TRUE) {
            await(self \notin waitP);
            put(self);
        }
    }

    process (c \in Consumers) {
c0:     while (TRUE) {
            await(self \notin waitC);
            get(self);
       }
    }
}
*)
\* BEGIN TRANSLATION (chksum(pcal) = "f5c8e8d0" /\ chksum(tla) = "191da49c")
VARIABLES 
    \* @type: Seq(PROC);
    buffer, 
    \* @type: Set(PROC);
    waitP, 
    \* @type: Set(PROC);
    waitC

(* define statement *)
isfull(b) == Len(b) = BufCapacity
isempty(b) == Len(b) = 0

\* @type: << Seq(PROC), Set(PROC), Set(PROC) >>;
vars == << buffer, waitP, waitC >>

ProcSet == (Producers) \cup (Consumers)

Init == (* Global variables *)
        /\ buffer = << >>
        /\ waitP = {}
        /\ waitC = {}

p(self) == /\ (self \notin waitP)
           /\ IF isfull(buffer)
                 THEN /\ IF self \in Producers
                            THEN /\ waitP' = (waitP \union {self})
                                 /\ waitC' = waitC
                            ELSE /\ waitC' = (waitC \union {self})
                                 /\ waitP' = waitP
                      /\ UNCHANGED buffer
                 ELSE /\ buffer' = Append(buffer, self)
                      /\ IF waitC # {}
                            THEN /\ \E t \in waitC:
                                      waitC' = waitC \ {t}
                            ELSE /\ TRUE
                                 /\ waitC' = waitC
                      /\ waitP' = waitP

c(self) == /\ (self \notin waitC)
           /\ IF isempty(buffer)
                 THEN /\ IF self \in Producers
                            THEN /\ waitP' = (waitP \union {self})
                                 /\ waitC' = waitC
                            ELSE /\ waitC' = (waitC \union {self})
                                 /\ waitP' = waitP
                      /\ UNCHANGED buffer
                 ELSE /\ buffer' = Tail(buffer)
                      /\ IF waitP # {}
                            THEN /\ \E t \in waitP:
                                      waitP' = waitP \ {t}
                            ELSE /\ TRUE
                                 /\ waitP' = waitP
                      /\ waitC' = waitC

Next == (\E self \in Producers: p(self))
           \/ (\E self \in Consumers: c(self))

Spec == Init /\ [][Next]_vars

\* END TRANSLATION 

Invariant == ~(waitP = Producers /\ waitC = Consumers)

=============================================================================
````

## The command line parameters used to run the tool

```
--config=BlockingQueue.cfg
```

## Log files

<details>

```
2023-09-28T15:34:49,306 [main] INFO  a.f.a.t.Tool\$ - # APALACHE version: 0.43.0 | build: 1db427e
2023-09-28T15:34:49,338 [main] INFO  a.f.a.i.p.o.OptionGroup\$ -   > BlockingQueue.cfg: Loading TLC configuration
2023-09-28T15:34:49,427 [main] INFO  a.f.a.i.p.o.OptionGroup\$ -   > Using inv predicate(s) Invariant from the TLC config
2023-09-28T15:34:49,429 [main] INFO  a.f.a.t.t.o.CheckCmd - Tuning: search.outputTraces=false
2023-09-28T15:34:49,678 [main] INFO  a.f.a.i.p.PassChainExecutor\$ - PASS #0: SanyParser
2023-09-28T15:34:50,008 [main] DEBUG a.f.a.i.p.PassChainExecutor\$ - PASS #0: SanyParser [OK]
2023-09-28T15:34:50,009 [main] INFO  a.f.a.i.p.PassChainExecutor\$ - PASS #1: TypeCheckerSnowcat
2023-09-28T15:34:50,009 [main] INFO  a.f.a.t.p.t.EtcTypeCheckerPassImpl -  > Running Snowcat .::.
2023-09-28T15:34:50,332 [main] INFO  a.f.a.t.p.t.EtcTypeCheckerPassImpl -  > Your types are purrfect!
2023-09-28T15:34:50,333 [main] INFO  a.f.a.t.p.t.EtcTypeCheckerPassImpl -  > All expressions are typed
2023-09-28T15:34:50,333 [main] DEBUG a.f.a.i.p.PassChainExecutor\$ - PASS #1: TypeCheckerSnowcat [OK]
2023-09-28T15:34:50,333 [main] INFO  a.f.a.i.p.PassChainExecutor\$ - PASS #2: ConfigurationPass
2023-09-28T15:34:50,349 [main] INFO  a.f.a.t.p.p.ConfigurationPassImpl -   > BlockingQueue.cfg: Using SPECIFICATION Spec
2023-09-28T15:34:50,351 [main] INFO  a.f.a.t.p.p.ConfigurationPassImpl -   > BlockingQueue.cfg: found INVARIANTS: Invariant
2023-09-28T15:34:50,353 [main] INFO  a.f.a.t.p.p.ConfigurationPassImpl -   > Set the initialization predicate to Init
2023-09-28T15:34:50,354 [main] INFO  a.f.a.t.p.p.ConfigurationPassImpl -   > Set the transition predicate to Next
2023-09-28T15:34:50,354 [main] INFO  a.f.a.t.p.p.ConfigurationPassImpl -   > Set the constant initialization predicate to CInit
2023-09-28T15:34:50,355 [main] INFO  a.f.a.t.p.p.ConfigurationPassImpl -   > Set an invariant to Invariant
2023-09-28T15:34:50,359 [main] DEBUG a.f.a.i.p.PassChainExecutor\$ - PASS #2: ConfigurationPass [OK]
2023-09-28T15:34:50,360 [main] INFO  a.f.a.i.p.PassChainExecutor\$ - PASS #3: DesugarerPass
2023-09-28T15:34:50,360 [main] INFO  a.f.a.t.p.p.DesugarerPassImpl -   > Desugaring...
2023-09-28T15:34:50,553 [main] DEBUG a.f.a.i.p.PassChainExecutor\$ - PASS #3: DesugarerPass [OK]
2023-09-28T15:34:50,553 [main] INFO  a.f.a.i.p.PassChainExecutor\$ - PASS #4: InlinePass
2023-09-28T15:34:50,554 [main] INFO  a.f.a.t.p.p.InlinePassImpl - Leaving only relevant operators: CInit, CInitPrimed, Init, InitPrimed, Invariant, Next
2023-09-28T15:34:50,582 [main] DEBUG a.f.a.i.p.PassChainExecutor\$ - PASS #4: InlinePass [OK]
2023-09-28T15:34:50,582 [main] INFO  a.f.a.i.p.PassChainExecutor\$ - PASS #5: TemporalPass
2023-09-28T15:34:50,582 [main] INFO  a.f.a.t.p.p.TemporalPassImpl -   > Rewriting temporal operators...
2023-09-28T15:34:50,583 [main] INFO  a.f.a.t.p.p.TemporalPassImpl -   > No temporal property specified, nothing to encode
2023-09-28T15:34:50,583 [main] DEBUG a.f.a.i.p.PassChainExecutor\$ - PASS #5: TemporalPass [OK]
2023-09-28T15:34:50,583 [main] INFO  a.f.a.i.p.PassChainExecutor\$ - PASS #6: InlinePass
2023-09-28T15:34:50,583 [main] INFO  a.f.a.t.p.p.InlinePassImpl - Leaving only relevant operators: CInit, CInitPrimed, Init, InitPrimed, Invariant, Next
2023-09-28T15:34:50,587 [main] DEBUG a.f.a.i.p.PassChainExecutor\$ - PASS #6: InlinePass [OK]
2023-09-28T15:34:50,587 [main] INFO  a.f.a.i.p.PassChainExecutor\$ - PASS #7: PrimingPass
2023-09-28T15:34:50,589 [main] INFO  a.f.a.t.p.a.PrimingPassImpl -   > Introducing CInitPrimed for CInit'
2023-09-28T15:34:50,593 [main] INFO  a.f.a.t.p.a.PrimingPassImpl -   > Introducing InitPrimed for Init'
2023-09-28T15:34:50,594 [main] DEBUG a.f.a.i.p.PassChainExecutor\$ - PASS #7: PrimingPass [OK]
2023-09-28T15:34:50,594 [main] INFO  a.f.a.i.p.PassChainExecutor\$ - PASS #8: VCGen
2023-09-28T15:34:50,595 [main] INFO  a.f.a.t.b.p.VCGenPassImpl -   > Producing verification conditions from the invariant Invariant
2023-09-28T15:34:50,602 [main] INFO  a.f.a.t.b.VCGenerator -   > VCGen produced 1 verification condition(s)
2023-09-28T15:34:50,603 [main] DEBUG a.f.a.i.p.PassChainExecutor\$ - PASS #8: VCGen [OK]
2023-09-28T15:34:50,604 [main] INFO  a.f.a.i.p.PassChainExecutor\$ - PASS #9: PreprocessingPass
2023-09-28T15:34:50,604 [main] INFO  a.f.a.t.p.p.PreproPassImpl -   > Before preprocessing: unique renaming
2023-09-28T15:34:50,609 [main] INFO  a.f.a.t.p.p.PreproPassImpl -  > Applying standard transformations:
2023-09-28T15:34:50,610 [main] INFO  a.f.a.t.p.p.PreproPassImpl -   > PrimePropagation
2023-09-28T15:34:50,612 [main] INFO  a.f.a.t.p.p.PreproPassImpl -   > Desugarer
2023-09-28T15:34:50,614 [main] INFO  a.f.a.t.p.p.PreproPassImpl -   > UniqueRenamer
2023-09-28T15:34:50,619 [main] INFO  a.f.a.t.p.p.PreproPassImpl -   > Normalizer
2023-09-28T15:34:50,622 [main] INFO  a.f.a.t.p.p.PreproPassImpl -   > Keramelizer
2023-09-28T15:34:50,628 [main] INFO  a.f.a.t.p.p.PreproPassImpl -   > After preprocessing: UniqueRenamer
2023-09-28T15:34:50,634 [main] DEBUG a.f.a.i.p.PassChainExecutor\$ - PASS #9: PreprocessingPass [OK]
2023-09-28T15:34:50,634 [main] INFO  a.f.a.i.p.PassChainExecutor\$ - PASS #10: TransitionFinderPass
2023-09-28T15:34:50,649 [main] INFO  a.f.a.t.p.a.TransitionPassImpl -   > Found 1 initializing transitions
2023-09-28T15:34:50,658 [main] INFO  a.f.a.t.p.a.TransitionPassImpl -   > Found 8 transitions
2023-09-28T15:34:50,659 [main] INFO  a.f.a.t.p.a.TransitionPassImpl -   > Found constant initializer CInit
2023-09-28T15:34:50,661 [main] INFO  a.f.a.t.p.a.TransitionPassImpl -   > Applying unique renaming
2023-09-28T15:34:50,667 [main] DEBUG a.f.a.i.p.PassChainExecutor\$ - PASS #10: TransitionFinderPass [OK]
2023-09-28T15:34:50,667 [main] INFO  a.f.a.i.p.PassChainExecutor\$ - PASS #11: OptimizationPass
2023-09-28T15:34:50,671 [main] INFO  a.f.a.t.p.p.OptPassImpl -  > Applying optimizations:
2023-09-28T15:34:50,672 [main] INFO  a.f.a.t.p.p.OptPassImpl -   > ConstSimplifier
2023-09-28T15:34:50,677 [main] INFO  a.f.a.t.p.p.OptPassImpl -   > ExprOptimizer
2023-09-28T15:34:50,681 [main] INFO  a.f.a.t.p.p.OptPassImpl -   > SetMembershipSimplifier
2023-09-28T15:34:50,683 [main] INFO  a.f.a.t.p.p.OptPassImpl -   > ConstSimplifier
2023-09-28T15:34:50,687 [main] DEBUG a.f.a.i.p.PassChainExecutor\$ - PASS #11: OptimizationPass [OK]
2023-09-28T15:34:50,687 [main] INFO  a.f.a.i.p.PassChainExecutor\$ - PASS #12: AnalysisPass
2023-09-28T15:34:50,689 [main] INFO  a.f.a.t.b.p.AnalysisPassImpl -  > Marking skolemizable existentials and sets to be expanded...
2023-09-28T15:34:50,690 [main] INFO  a.f.a.t.b.p.AnalysisPassImpl -   > Skolemization
2023-09-28T15:34:50,691 [main] INFO  a.f.a.t.b.p.AnalysisPassImpl -   > Expansion
2023-09-28T15:34:50,692 [main] INFO  a.f.a.t.b.p.AnalysisPassImpl -   > Remove unused let-in defs
2023-09-28T15:34:50,696 [main] INFO  a.f.a.t.b.p.AnalysisPassImpl -  > Running analyzers...
2023-09-28T15:34:50,700 [main] INFO  a.f.a.t.b.p.AnalysisPassImpl -   > Introduced expression grades
2023-09-28T15:34:50,701 [main] DEBUG a.f.a.i.p.PassChainExecutor\$ - PASS #12: AnalysisPass [OK]
2023-09-28T15:34:50,701 [main] INFO  a.f.a.i.p.PassChainExecutor\$ - PASS #13: BoundedChecker
2023-09-28T15:34:50,714 [main] DEBUG a.f.a.t.b.s.Z3SolverContext - Creating Z3 solver context 0
2023-09-28T15:34:50,973 [main] DEBUG a.f.a.t.b.t.TransitionExecutorImpl - Initializing CONSTANTS
2023-09-28T15:34:51,027 [main] ERROR a.f.a.t.Tool\$ - Unhandled exception
java.lang.AssertionError: assertion failed
	at scala.Predef\$.assert(Predef.scala:264)
	at at.forsyte.apalache.tla.bmcmt.LazyEquality.mkSetEq(LazyEquality.scala:251)
	at at.forsyte.apalache.tla.bmcmt.LazyEquality.cacheOneEqConstraint(LazyEquality.scala:146)
	at at.forsyte.apalache.tla.bmcmt.rules.EqRule.apply(EqRule.scala:55)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.rewriteOnce(SymbStateRewriterImpl.scala:355)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.doRecursive\$1(SymbStateRewriterImpl.scala:388)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.rewriteUntilDone(SymbStateRewriterImpl.scala:422)
	at at.forsyte.apalache.tla.bmcmt.rules.NegRule.apply(NegRule.scala:27)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.rewriteOnce(SymbStateRewriterImpl.scala:355)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.doRecursive\$1(SymbStateRewriterImpl.scala:388)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.rewriteUntilDone(SymbStateRewriterImpl.scala:422)
	at at.forsyte.apalache.tla.bmcmt.rules.AndRule.mapArg\$1(AndRule.scala:62)
	at at.forsyte.apalache.tla.bmcmt.rules.AndRule.\$anonfun\$apply\$2(AndRule.scala:66)
	at scala.collection.immutable.List.map(List.scala:246)
	at scala.collection.immutable.List.map(List.scala:79)
	at at.forsyte.apalache.tla.bmcmt.rules.AndRule.apply(AndRule.scala:66)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.rewriteOnce(SymbStateRewriterImpl.scala:355)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.doRecursive\$1(SymbStateRewriterImpl.scala:388)
	at at.forsyte.apalache.tla.bmcmt.SymbStateRewriterImpl.rewriteUntilDone(SymbStateRewriterImpl.scala:422)
	at at.forsyte.apalache.tla.bmcmt.trex.TransitionExecutorImpl.assertState(TransitionExecutorImpl.scala:216)
	at at.forsyte.apalache.tla.bmcmt.trex.FilteredTransitionExecutor.assertState(FilteredTransitionExecutor.scala:123)
	at at.forsyte.apalache.tla.bmcmt.trex.ConstrainedTransitionExecutor.assertState(ConstrainedTransitionExecutor.scala:111)
	at at.forsyte.apalache.tla.bmcmt.SeqModelChecker.\$anonfun\$run\$1(SeqModelChecker.scala:54)
	at at.forsyte.apalache.tla.bmcmt.SeqModelChecker.\$anonfun\$run\$1\$adapted(SeqModelChecker.scala:53)
	at scala.collection.immutable.List.foreach(List.scala:333)
	at at.forsyte.apalache.tla.bmcmt.SeqModelChecker.run(SeqModelChecker.scala:53)
	at at.forsyte.apalache.tla.bmcmt.passes.BoundedCheckerPassImpl.runIncrementalChecker(BoundedCheckerPassImpl.scala:135)
	at at.forsyte.apalache.tla.bmcmt.passes.BoundedCheckerPassImpl.execute(BoundedCheckerPassImpl.scala:88)
	at at.forsyte.apalache.infra.passes.PassChainExecutor\$.exec(PassChainExecutor.scala:64)
	at at.forsyte.apalache.infra.passes.PassChainExecutor\$.\$anonfun\$runPassOnModule\$3(PassChainExecutor.scala:53)
	at scala.util.Either.flatMap(Either.scala:352)
	at at.forsyte.apalache.infra.passes.PassChainExecutor\$.\$anonfun\$runPassOnModule\$1(PassChainExecutor.scala:51)
	at scala.collection.LinearSeqOps.foldLeft(LinearSeq.scala:183)
	at scala.collection.LinearSeqOps.foldLeft\$(LinearSeq.scala:179)
	at scala.collection.immutable.List.foldLeft(List.scala:79)
	at at.forsyte.apalache.infra.passes.PassChainExecutor\$.runOnPasses(PassChainExecutor.scala:44)
	at at.forsyte.apalache.infra.passes.PassChainExecutor\$.run(PassChainExecutor.scala:34)
	at at.forsyte.apalache.tla.tooling.opt.CheckCmd.run(CheckCmd.scala:132)
	at at.forsyte.apalache.tla.Tool\$.runCommand(Tool.scala:138)
	at at.forsyte.apalache.tla.Tool\$.run(Tool.scala:118)
	at at.forsyte.apalache.tla.Tool\$.main(Tool.scala:40)
	at at.forsyte.apalache.tla.Tool.main(Tool.scala)
```
</details>

## System information

- Apalache version: `0.43.0 build 1db427e`
- OS: `Mac OS X`
- JDK version: `17.0.2`

## Triage checklist (for maintainers)

- [ ] Reproduce the bug on the main development branch.
- [ ] Add the issue to the apalache GitHub project.
- [ ] If the bug is high impact, ensure someone available is assigned to fix it.

[report.zip](https://github.com/informalsystems/apalache/files/12750177/report.zip)

### PR DESCRIPTION
**Pull request by Kukovec**
- [x] Tests added for any new code
- [x] Ran `make fmt-fix` (or had formatting run automatically on all files edited)
- [ ] Documentation added for any new functionality
- [ ] [Entries added to `./unreleased/`][changelog format] for any new functionality

This PR adds an additional type validation check to any CONSTANT overwritten from a .cfg file. Since the typechecking pass precedes the configuration pass, it was otherwise possible to introduce type-inconsistencies, by overwriting a constant declared with type `A` with a value of type `B`, which would only fail at the cell-introduction level, with cryptic error messages.

[72aaea0](https://github.com/informalsystems/apalache/pull/2757/commits/72aaea0723dcb3eb3e34715f2de6e6582d35e9bb) adds support for literals of uninterpreted types within .cfg files (+ tests and some cleanup to use the typed builder smoothly)

closes #2750

[changelog format]: https://github.com/informalsystems/apalache/blob/main/CONTRIBUTING.md#how-to-record-a-change

### CODE REVIEW

File: mod-infra/src/main/scala/at/forsyte/apalache/infra/tlc/config/TlcConfig.scala [hunk 1]
```diff
@@ -64,11 +64,11 @@ object ConfigModelValue {
  *   the name of a model value
  */
 case class ConfigModelValue(name: String) extends ConfigConstExpr {
```
Comment (konnov) [inline mod-infra/src/main/scala/at/forsyte/apalache/infra/tlc/config/TlcConfig.scala:66]: What is the difference between `ConfigModelValue` and `ConfigStrValue` now? It looks like `ConfigStrValue` handles model values too.

**Review by konnov — APPROVED** [review]
Looks good! I just left one comment. Thanks for fixing this!

File: test/tla/Bug2750.cfg [hunk 2]
```diff
@@ -0,0 +1,9 @@
+\* Add statements after this line.
+SPECIFICATION Spec
+CONSTANTS 
+    Producers = {p1,p2}
+    Consumers = {c1,c2}
+    BufCapacity = 1
+\* generate dot file: add "-dump dot trans.dot" to TLC options
+\* (=> extension settings)
```
Comment (thpani) [inline test/tla/Bug2750.cfg:8]: I was confused at first, how this is relevant for the regression test.
Then I realized it's not. Can we remove it?
```suggestion
```

**Review by thpani — APPROVED** [review]
Looks great!

Amazing that it now works with uninterpreted types

File: mod-infra/src/main/scala/at/forsyte/apalache/infra/tlc/config/TlcConfig.scala (hunk 1 above)
Comment (Kukovec) [reply mod-infra/src/main/scala/at/forsyte/apalache/infra/tlc/config/TlcConfig.scala:66]: `ConfigModelValue` is what you get when TLC model values are parsed (e.g. `a` and `b` in `Producers = {a,b}`).
`ConfigStrValue` is what you get when a `"`-value is parsed (e.g. `"abc"` or `"1_OF_X"`)

File: test/tla/Bug2750.cfg (hunk 2 above)
Comment (Kukovec) [reply test/tla/Bug2750.cfg:8]: I just copied the files as they were submitted in #2750, and only renamed them. We can axe this if you like.

**Pull request merged**